// zombie/lib/storage.js
// See [Web Storage](http://dev.w3.org/html5/webstorage/)
'use strict';

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _get = require('babel-runtime/helpers/get')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _slicedToArray = require('babel-runtime/helpers/sliced-to-array')['default'];

var _core = require('babel-runtime/core-js')['default'];

var DOM = require('./dom');

// Implementation of the StorageEvent.

var StorageEvent = (function (_DOM$Event) {
  function StorageEvent(storage, url, key, oldValue, newValue) {
    _classCallCheck(this, StorageEvent);

    _get(_core.Object.getPrototypeOf(StorageEvent.prototype), 'constructor', this).call(this, 'storage');
    this._storage = storage;
    this._url = url;
    this._key = key;
    this._oldValue = oldValue;
    this._newValue = newValue;
  }

  _inherits(StorageEvent, _DOM$Event);

  _createClass(StorageEvent, [{
    key: 'url',
    get: function () {
      return this._url;
    }
  }, {
    key: 'storageArea',
    get: function () {
      return this._storage;
    }
  }, {
    key: 'key',
    get: function () {
      return this._key;
    }
  }, {
    key: 'oldValue',
    get: function () {
      return this._oldValue;
    }
  }, {
    key: 'newValue',
    get: function () {
      return this._newValue;
    }
  }]);

  return StorageEvent;
})(DOM.Event);

// Storage area. The storage area is shared by multiple documents of the same
// origin. For session storage, they must also share the same browsing context.

var StorageArea = (function () {
  function StorageArea() {
    _classCallCheck(this, StorageArea);

    this._items = {};
    this._storages = [];
  }

  _createClass(StorageArea, [{
    key: '_fire',

    // Fire a storage event. Fire in all documents that share this storage area,
    // except for the source document.
    value: function _fire(source, key, oldValue, newValue) {
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = _core.getIterator(this._storages), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var _step$value = _slicedToArray(_step.value, 2);

          var storage = _step$value[0];
          var _window = _step$value[1];

          if (storage === source) continue;
          var _event = new StorageEvent(storage, _window.location.href, key, oldValue, newValue);
          _window.dispatchEvent(_event);
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator['return']) {
            _iterator['return']();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }
  }, {
    key: 'length',

    // Return number of key/value pairs.
    get: function () {
      return _core.Object.keys(this._items).length;
    }
  }, {
    key: 'key',

    // Get key by ordinal position.
    value: function key(index) {
      return _core.Object.keys(this._items)[index];
    }
  }, {
    key: 'get',

    // Get value from key
    value: function get(key) {
      return this._items[key] || null;
    }
  }, {
    key: 'set',

    // Set the value of a key. We also need the source storage (so we don't send
    // it a storage event).
    value: function set(source, key, value) {
      var oldValue = this._items[key];
      this._items[key] = value;
      this._fire(source, key, oldValue, value);
    }
  }, {
    key: 'remove',

    // Remove the value at the key. We also need source storage (see set above).
    value: function remove(source, key) {
      var oldValue = this._items[key];
      delete this._items[key];
      this._fire(source, key, oldValue);
    }
  }, {
    key: 'clear',

    // Remove all values. We also need source storage (see set above).
    value: function clear(source) {
      this._items = {};
      this._fire(source);
    }
  }, {
    key: 'pairs',
    get: function () {
      var _this3 = this;

      return _core.Object.keys(this._items).map(function (key) {
        return [key, _this3._items[key]];
      });
    }
  }, {
    key: 'toString',
    value: function toString() {
      var _this4 = this;

      return _core.Object.keys(this._items).map(function (key) {
        return '' + key + ' = ' + _this4._items[key];
      }).join('\n');
    }
  }, {
    key: 'associate',

    // Associate local/sessionStorage and window with this storage area. Used when firing events.
    value: function associate(storage, window) {
      this._storages.push([storage, window]);
    }
  }]);

  return StorageArea;
})();

// Implementation of the Storage interface, used by local and session storage.

var Storage = (function () {
  function Storage(area) {
    _classCallCheck(this, Storage);

    this._area = area;
  }

  _createClass(Storage, [{
    key: 'length',

    // ### storage.length => Number
    //
    // Returns the number of key/value pairs in this storage.
    get: function () {
      return this._area.length;
    }
  }, {
    key: 'key',

    // ### storage.key(index) => String
    //
    // Returns the key at this position.
    value: function key(index) {
      return this._area.key(index);
    }
  }, {
    key: 'getItem',

    // ### storage.getItem(key) => Object
    //
    // Returns item by key.
    value: function getItem(key) {
      return this._area.get(key.toString());
    }
  }, {
    key: 'setItem',

    // ### storage.setItem(key, Object)
    //
    // Add item or change value of existing item.
    value: function setItem(key, value) {
      this._area.set(this, key.toString(), value);
    }
  }, {
    key: 'removeItem',

    // ### storage.removeItem(key)
    //
    // Remove item.
    value: function removeItem(key) {
      this._area.remove(this, key.toString());
    }
  }, {
    key: 'clear',

    // ### storage.clear()
    //
    // Remove all items.
    value: function clear() {
      this._area.clear(this);
    }
  }, {
    key: 'dump',

    // Dump to a string, useful for debugging.
    value: function dump() {
      var output = arguments[0] === undefined ? process.stdout : arguments[0];

      return this._area.dump(output);
    }
  }]);

  return Storage;
})();

// Combined local/session storage.

var Storages = (function () {
  function Storages() {
    _classCallCheck(this, Storages);

    this._locals = {};
    this._sessions = {};
  }

  _createClass(Storages, [{
    key: 'local',

    // Return local Storage based on the document origin (hostname/port).
    value: function local(host) {
      if (!this._locals[host]) this._locals[host] = new StorageArea();
      return new Storage(this._locals[host]);
    }
  }, {
    key: 'session',

    // Return session Storage based on the document origin (hostname/port).
    value: function session(host) {
      if (!this._sessions[host]) this._sessions[host] = new StorageArea();
      return new Storage(this._sessions[host]);
    }
  }, {
    key: 'extend',

    // Extend window with local/session storage support.
    value: function extend(window) {
      var storages = this;
      window.StorageEvent = StorageEvent;
      _core.Object.defineProperties(window, {
        localStorage: {
          get: function get() {
            var document = this.document;

            if (!document._localStorage) document._localStorage = storages.local(document.location.host);
            return document._localStorage;
          }
        },

        sessionStorage: {
          get: function get() {
            var document = this.document;

            if (!document._sessionStorage) document._sessionStorage = storages.session(document.location.host);
            return document._sessionStorage;
          }
        }
      });
    }
  }, {
    key: 'dump',

    // Used to dump state to console (debuggin)
    value: function dump() {
      var output = arguments[0] === undefined ? process.stdout : arguments[0];
      var _iteratorNormalCompletion2 = true;
      var _didIteratorError2 = false;
      var _iteratorError2 = undefined;

      try {
        for (var _iterator2 = _core.getIterator(this._locals), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          var domain = _step2.value;

          var area = this._locals[domain];
          output.write('' + domain + ' local:\n');
          var _iteratorNormalCompletion4 = true;
          var _didIteratorError4 = false;
          var _iteratorError4 = undefined;

          try {
            for (var _iterator4 = _core.getIterator(area.pairs), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
              var _step4$value = _slicedToArray(_step4.value, 2);

              var _name = _step4$value[0];
              var value = _step4$value[1];

              output.write('  ' + _name + ' = ' + value + '\n');
            }
          } catch (err) {
            _didIteratorError4 = true;
            _iteratorError4 = err;
          } finally {
            try {
              if (!_iteratorNormalCompletion4 && _iterator4['return']) {
                _iterator4['return']();
              }
            } finally {
              if (_didIteratorError4) {
                throw _iteratorError4;
              }
            }
          }
        }
      } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion2 && _iterator2['return']) {
            _iterator2['return']();
          }
        } finally {
          if (_didIteratorError2) {
            throw _iteratorError2;
          }
        }
      }

      var _iteratorNormalCompletion3 = true;
      var _didIteratorError3 = false;
      var _iteratorError3 = undefined;

      try {
        for (var _iterator3 = _core.getIterator(this._sessions), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
          var domain = _step3.value;

          var area = this._sessions[domain];
          output.push('' + domain + ' session:\n');
          var _iteratorNormalCompletion5 = true;
          var _didIteratorError5 = false;
          var _iteratorError5 = undefined;

          try {
            for (var _iterator5 = _core.getIterator(area.pairs), _step5; !(_iteratorNormalCompletion5 = (_step5 = _iterator5.next()).done); _iteratorNormalCompletion5 = true) {
              var _step5$value = _slicedToArray(_step5.value, 2);

              var _name2 = _step5$value[0];
              var value = _step5$value[1];

              output.write('  ' + _name2 + ' = ' + value + '\n');
            }
          } catch (err) {
            _didIteratorError5 = true;
            _iteratorError5 = err;
          } finally {
            try {
              if (!_iteratorNormalCompletion5 && _iterator5['return']) {
                _iterator5['return']();
              }
            } finally {
              if (_didIteratorError5) {
                throw _iteratorError5;
              }
            }
          }
        }
      } catch (err) {
        _didIteratorError3 = true;
        _iteratorError3 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion3 && _iterator3['return']) {
            _iterator3['return']();
          }
        } finally {
          if (_didIteratorError3) {
            throw _iteratorError3;
          }
        }
      }
    }
  }, {
    key: 'save',

    // browser.saveStorage uses this
    value: function save() {
      var serialized = ['# Saved on ' + new Date().toISOString()];
      var _iteratorNormalCompletion6 = true;
      var _didIteratorError6 = false;
      var _iteratorError6 = undefined;

      try {
        for (var _iterator6 = _core.getIterator(this._locals), _step6; !(_iteratorNormalCompletion6 = (_step6 = _iterator6.next()).done); _iteratorNormalCompletion6 = true) {
          var domain = _step6.value;

          var area = this._locals[domain];
          var pairs = area.pairs;
          if (pairs.length) {
            serialized.push('' + domain + ' local:');
            var _iteratorNormalCompletion8 = true;
            var _didIteratorError8 = false;
            var _iteratorError8 = undefined;

            try {
              for (var _iterator8 = _core.getIterator(area.pairs), _step8; !(_iteratorNormalCompletion8 = (_step8 = _iterator8.next()).done); _iteratorNormalCompletion8 = true) {
                var _step8$value = _slicedToArray(_step8.value, 2);

                var _name3 = _step8$value[0];
                var value = _step8$value[1];

                serialized.push('  ' + escape(_name3) + ' = ' + escape(value));
              }
            } catch (err) {
              _didIteratorError8 = true;
              _iteratorError8 = err;
            } finally {
              try {
                if (!_iteratorNormalCompletion8 && _iterator8['return']) {
                  _iterator8['return']();
                }
              } finally {
                if (_didIteratorError8) {
                  throw _iteratorError8;
                }
              }
            }
          }
        }
      } catch (err) {
        _didIteratorError6 = true;
        _iteratorError6 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion6 && _iterator6['return']) {
            _iterator6['return']();
          }
        } finally {
          if (_didIteratorError6) {
            throw _iteratorError6;
          }
        }
      }

      var _iteratorNormalCompletion7 = true;
      var _didIteratorError7 = false;
      var _iteratorError7 = undefined;

      try {
        for (var _iterator7 = _core.getIterator(this._sessions), _step7; !(_iteratorNormalCompletion7 = (_step7 = _iterator7.next()).done); _iteratorNormalCompletion7 = true) {
          var domain = _step7.value;

          var area = this._sessions[domain];
          var pairs = area.pairs;
          if (pairs.length) {
            serialized.push('' + domain + ' session:');
            var _iteratorNormalCompletion9 = true;
            var _didIteratorError9 = false;
            var _iteratorError9 = undefined;

            try {
              for (var _iterator9 = _core.getIterator(area.pairs), _step9; !(_iteratorNormalCompletion9 = (_step9 = _iterator9.next()).done); _iteratorNormalCompletion9 = true) {
                var _step9$value = _slicedToArray(_step9.value, 2);

                var _name4 = _step9$value[0];
                var value = _step9$value[1];

                serialized.push('  ' + escape(_name4) + ' = ' + escape(value));
              }
            } catch (err) {
              _didIteratorError9 = true;
              _iteratorError9 = err;
            } finally {
              try {
                if (!_iteratorNormalCompletion9 && _iterator9['return']) {
                  _iterator9['return']();
                }
              } finally {
                if (_didIteratorError9) {
                  throw _iteratorError9;
                }
              }
            }
          }
        }
      } catch (err) {
        _didIteratorError7 = true;
        _iteratorError7 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion7 && _iterator7['return']) {
            _iterator7['return']();
          }
        } finally {
          if (_didIteratorError7) {
            throw _iteratorError7;
          }
        }
      }

      return serialized.join('\n') + '\n';
    }
  }, {
    key: 'load',

    // browser.loadStorage uses this
    value: function load(serialized) {
      var storage = null;
      var _iteratorNormalCompletion10 = true;
      var _didIteratorError10 = false;
      var _iteratorError10 = undefined;

      try {
        for (var _iterator10 = _core.getIterator(serialized.split(/\n+/)), _step10; !(_iteratorNormalCompletion10 = (_step10 = _iterator10.next()).done); _iteratorNormalCompletion10 = true) {
          var item = _step10.value;

          if (item[0] === '#' || item === '') continue;
          if (item[0] === ' ') {
            var _item$split = item.split('=');

            var _item$split2 = _slicedToArray(_item$split, 2);

            var key = _item$split2[0];
            var value = _item$split2[1];

            if (storage) storage.setItem(unescape(key.trim()), unescape(value.trim()));else throw new Error('Must specify storage type using local: or session:');
          } else {
            var _item$split3 = item.split(' ');

            var _item$split32 = _slicedToArray(_item$split3, 2);

            var domain = _item$split32[0];
            var type = _item$split32[1];

            if (type === 'local:') storage = this.local(domain);else if (type === 'session:') storage = this.session(domain);else throw new Error('Unkown storage type ' + type);
          }
        }
      } catch (err) {
        _didIteratorError10 = true;
        _iteratorError10 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion10 && _iterator10['return']) {
            _iterator10['return']();
          }
        } finally {
          if (_didIteratorError10) {
            throw _iteratorError10;
          }
        }
      }
    }
  }]);

  return Storages;
})();

module.exports = Storages;
